string/string.cc

```diff
@@ -1,12 +1,154 @@
 #include "string.h"
 
-String::String()
+#include <algorithm>
+#include <iterator>
+#include <memory>
+#include <string>
+#include <cstring>
+
+const char* EMPTY = "";
+
+String::String() = default;
+
+String::~String() = default;
+
+String::String(String&& o) = default;
+
+String::String(const String& o): String(o.begin(), o.end()) {}
+
+String::String(const char* s): String(s, s + std::strlen(s)) {}
+
+template String::String(std::string::iterator first, std::string::iterator last);
+
+template<typename It> String::String(It first, It last)
+{
+	mSize = std::distance(first, last);
+	mValue.reset(new char[mSize + 1]);
+
+	std::copy_n(first, mSize, begin());
+
+	mValue[mSize + 1] = '\0';
+}
+
+// I don't see any other effective way to concatenate for the given API.
+String::String(const String& a, const String& b)
```

Comment:
And string buffer doesn't contain a, uhm, a buffer to which the strings have to be copied?

For 2 strings there is no reasonable optimization. You may want to try to optimize an expression like this:

    String s = a + b + c + d;

by allocating buffer only once. Look for 'expression templates'.

Also, you could optimize operations like this:

    s += x;

by having a pre-allocated buffer in s bigger then needed. This would incur some initial cost, and would allow for avoiding allocation in some cases.